Polishing Secure Aggregation

aunell opened the pull request:
This is a PR for all of the details that we want to fix for the secure aggregation merge. A complete list of tasks that we will be working on/implementing can be found here: https://github.com/epfml/disco/issues/9.

Commit:
minimumReadyPeers now task dependent

Commit:
eslint

Commit:
made default values for trainingInformation fields, add readme

martinjaggi commented on README.md:
```diff
@@ -30,6 +30,7 @@ ___
 :question: **DISCO TECHNOLOGY** 
 - DISCO supports arbitrary deep learning tasks and model architectures, via [TF.js](https://www.tensorflow.org/js)
 - :sparkles: relies on [peer2peer](https://peerjs.com/) communication
+- Learn more about Secure Aggregation for privacy preservation in decentralized training [HERE](discojs/secAggREADME.md)
```
for privacy-respecting training

martinjaggi commented on discojs/secAggREADME.md:
```diff
@@ -0,0 +1,20 @@
+# Secure Multiparty Aggregation for Decentralized Training
```
skip dec.., will be federated as well soon

martinjaggi commented on discojs/secAggREADME.md:
```diff
@@ -0,0 +1,20 @@
+
+### Algorithm description
+Additive sharing (you can read more about this algorithm [HERE](https://www.geeksforgeeks.org/additive-secret-sharing-and-share-proactivization-using-python/?ref=rp)
+) is implemented with each client's gradients when the gradient is sent to all other clients connected in this round of multiparty training. The algorithm has four steps:
```
for generality, say model updates / models (can be both in the future), instead of gradients

martinjaggi commented on discojs/secAggREADME.md:
```diff
@@ -0,0 +1,20 @@
+### Algorithm description
+Additive sharing (you can read more about this algorithm [HERE](https://www.geeksforgeeks.org/additive-secret-sharing-and-share-proactivization-using-python/?ref=rp)
+) is implemented with each client's gradients when the gradient is sent to all other clients connected in this round of multiparty training. The algorithm has four steps:
+1. Each client will send a "ready message" to the signaling server, indicating that they are done with training and ready to be connected to other clients in order to aggregate weights. Each client pauses aggregation until they have received a list of peers to connect to.
```
not done with training yet, but with the round of local updates

martinjaggi commented on discojs/secAggREADME.md:
```diff
@@ -0,0 +1,20 @@
+Additive sharing (you can read more about this algorithm [HERE](https://www.geeksforgeeks.org/additive-secret-sharing-and-share-proactivization-using-python/?ref=rp)
+) is implemented with each client's gradients when the gradient is sent to all other clients connected in this round of multiparty training. The algorithm has four steps:
+1. Each client will send a "ready message" to the signaling server, indicating that they are done with training and ready to be connected to other clients in order to aggregate weights. Each client pauses aggregation until they have received a list of peers to connect to.
+2. Once the client has a list of peers, they will generate *n* shares, where _n_ is the number of clients (including themself) who will receive a share in order to be a part of this round of secure aggregation. Each client will receive 1 share from each other client. Each client will then have _n_ shares.
```
secret share, generated out of the secret sth value (model or model update)

martinjaggi replied:
say intuitively what this is. the true value additively masked with some secret noise

Grim-bot committed:
Merge branch 'develop' into 9-userInput-Alyssa-Felix

martinjaggi commented on discojs/secAggREADME.md:
```diff
@@ -0,0 +1,20 @@
+1. Each client will send a "ready message" to the signaling server, indicating that they are done with training and ready to be connected to other clients in order to aggregate weights. Each client pauses aggregation until they have received a list of peers to connect to.
+2. Once the client has a list of peers, they will generate *n* shares, where _n_ is the number of clients (including themself) who will receive a share in order to be a part of this round of secure aggregation. Each client will receive 1 share from each other client. Each client will then have _n_ shares.
+3. The client waits until they receive all _n_ shares (they are connected to themself as a peer, so they will receive a share from themself, as well), and when they do, they sum their shares together to create a _partial sum_ of all gradients. The client will then send this _partial sum_ to all peers that they are connected to.
+4. The client waits until they receive _n_ partial sums. When they receive all sums, they aggregate these sums and return the final gradient updates.
```
aggregate here means?

martinjaggi commented on discojs/secAggREADME.md:
```diff
@@ -0,0 +1,20 @@
+3. The client waits until they receive all _n_ shares (they are connected to themself as a peer, so they will receive a share from themself, as well), and when they do, they sum their shares together to create a _partial sum_ of all gradients. The client will then send this _partial sum_ to all peers that they are connected to.
+4. The client waits until they receive _n_ partial sums. When they receive all sums, they aggregate these sums and return the final gradient updates.
+### Privacy Guarantees
+Currently, if the gradient value exceeds an absolute value of 50, privacy of gradient values is sacrificed for precision of smaller gradient values. Additionally, it is
```
first mention the guarantee, input privacy. at the end add a small warning for model values >100 say

martinjaggi commented on discojs/secAggREADME.md:
```diff
@@ -0,0 +1,20 @@
+gradient update.
+
+### User Input
+The user can specify how many clients they want to be connected to before secure multiparty aggregation commences. This minimum number must be the same
```
in the task definition, link to readme for tasks

martinjaggi commented on discojs/src/client/decentralized_base.ts:
```diff
@@ -16,13 +17,22 @@ const TICK = 100
+  ) {
+    super(url, task)
+    this.minimumReadyPeers = this.task.trainingInformation?.minimumReadyPeers ?? 3
+    this.noiseMagnitude = this.task.trainingInformation?.noiseMagnitude ?? 50
```
rename so it's not confused with the DP noise magnitude

martinjaggi commented on discojs/src/secret_shares.ts:
```diff
@@ -47,40 +47,29 @@ export function lastShare (currentShares: Weights[], secret: Weights): Weights {
 }
 
-export function generateAllShares (secret: Weights, nParticipants: number): List<Weights> {
+export function generateAllShares (secret: Weights, nParticipants: number, noiseMagnitude: number): List<Weights> {
```
rename as above

martinjaggi commented on server/tests/end_to_end_decentralized.ts:
```diff
@@ -166,10 +166,10 @@ describe('end to end', function () {
-    makeClientSecure([3, 3, 3])
-    makeClientSecure([4, 5, 6])
-    return await makeClientSecure([11, 13, 30])
+    // expected --> [.002, 7, 27]
```
add some comment here if the array is over model dimension = 3 or number of clients =3

aunell replied:
I don't understand-- add a comment saying what about the model dimension or client numbers? currently the model doesn't undergo any training in this test -- it is just testing the communication of updated weights between clients.

Commit:
updated readme and renamed maximum random value parameter

Grim-bot committed:
Link to SECAGG.md from ARCHITECTURE.md

Commit:
updated secagg read me

Commit:
added to secagg read me, added comments and docstrings

martinjaggi commented:
here and in `makeClientSecure`, please add a comment on what the array means, what it ranges over, just for full clarity (otherwise people might easily confuse model dimension with number of clients). maybe also in some other places in the code. this is just for understanding&documentation, not a functional issue

martinjaggi committed:
minor polishing

Commit:
updated comments for clarity and readme

Commit:
merge conflict resolved

Commit:
clarify end to end test conventions, add dtype to tf.randomUniform()

Commit:
updated message structure, eliminated use of 'any' type, fixed all eslint and ts errors

Commit:
eslint

martinjaggi commented on discojs/src/client/decentralized/secret_shares.ts:
```diff
@@ -19,70 +19,62 @@ export function subtractWeights (w1: Weights, w2: Weights): Weights {
+/*
+generate a cryptographically secure random number with a maximum value of maxShareValue
+ */
+export function generateRandomNumber (maxShareValue: number): number {
```
we probably don't use this one anymore, as it's int not float. also is it signed or not (can be negative?)

aunell replied:
Does it matter if it's int if we are using it as a seed? It is not signed. When we are using random numbers to populate the generated shares, we are using signed floats within the range (-maxShareValue, maxShareValue).

Commit:
clarify seed vs random number generation for shares

Commit:
eslint

Commit:
simplified messaging interface in decentralized base

Commit:
delete testing file in discojs, update secret_shares.spec to work with console log statements

Grim-bot committed:
Minor edits to SECAGG and ARCHITECTURE

Grim-bot committed:
Fix typo

Commit:
updated testing strategy, made new file which has function to assert weight equality within epsilon

Commit:
eslint, need to fix functions that interact with updates (doesn't properly handle nested updates in secret shares or in our makeWeights function)

Grim-bot committed:
Proofread SECAGG.md, change to PRIVACY.md

- Changes SECAGG.md to PRIVACY.md, which discusses differential privacy
as well as secure update aggregation.
- Reworks the description of our secure aggregation implementation.

Grim-bot committed:
Merge branch '9-userInput-Alyssa-Felix' of github.com:epfml/disco into 9-userInput-Alyssa-Felix

martinjaggi committed:
minor

Grim-bot committed:
Reinstate the section title *Privacy guarantee*

Grim-bot committed:
Comment on possible but pointless efficiency gain.

Grim-bot committed:
Merge branch '9-userInput-Alyssa-Felix' of github.com:epfml/disco into 9-userInput-Alyssa-Felix

martinjaggi commented on discojs/src/client/decentralized/secret_shares.ts:
```diff
@@ -19,70 +20,58 @@ export function subtractWeights (w1: Weights, w2: Weights): Weights {
+ */
+export function generateRandomShare (secret: Weights, maxShareValue: number): Weights {
   const share: Weights = []
+  const seed: number = crypto.randomInt(maxSeed)
```
yes, much clearer now

martinjaggi committed:
Merge branch 'develop' into 9-userInput-Alyssa-Felix

martinjaggi committed:
trying to fix test

martinjaggi committed:
DP and secagg documentation

martinjaggi committed:
fix link to privacy docs

martinjaggi committed:
types import

martinjaggi committed:
Merge branch 'develop' into 9-userInput-Alyssa-Felix
